### PR TITLE
Make the message for async init not completed a lot clearer.

### PR DESCRIPTION
**Pull request by Sam-Kramer**
## General
**Before this PR**:
Occasionally users of the library would see this message, and not sure what to do or what it means.

**After this PR**:

==COMMIT_MSG==
Improved NotInitializedException exception message
==COMMIT_MSG==

**Priority**:
P0

**Concerns / possible downsides (what feedback would you like?)**:
- Is the exception message clear?

**Is documentation needed?**:
No

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**: 
It should be OK? Constructor is the same, message is slightly different. Safe log class should only be used if the instance of the object is SafeLoggable, which should be at runtime... 

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:
No

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:
Yes

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:
No

**Does this PR need a schema migration?**
No

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:
N/A

**What was existing testing like? What have you done to improve it?**:
N/A

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:
N/A

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:
N/A

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:
Logs

**Has the safety of all log arguments been decided correctly?**:
Yes

**Will this change significantly affect our spending on metrics or logs?**:
No

**How would I tell that this PR does not work in production? (monitors, etc.)**:
Logs!

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:
Rollback

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:
No

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:
No

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:
No

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-commons/src/main/java/com/palantir/exception/NotInitializedException.java [hunk 1]
```diff
@@ -16,18 +16,40 @@
...
 
 import com.palantir.common.exception.AtlasDbDependencyException;
+import com.palantir.logsafe.Arg;
+import com.palantir.logsafe.Safe;
 import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.SafeLoggable;
+
+import java.util.List;
+
+public class NotInitializedException extends AtlasDbDependencyException implements SafeLoggable {
```
Comment (Sam-Kramer) [inline atlasdb-commons/src/main/java/com/palantir/exception/NotInitializedException.java:25]: https://github.com/palantir/atlasdb/pull/4369/files we did previously, which should make this OK from an ABI break perspective. Also constructor args are the same, and I'd imagine `SafeLoggable` does an `instanceof` call, which hopefully is determined at runtime?

File: atlasdb-commons/src/main/java/com/palantir/exception/NotInitializedException.java [hunk 2]
```diff
@@ -16,18 +16,39 @@
...
+    private final String objectNotInitialized;
 
-public class NotInitializedException extends AtlasDbDependencyException {
     public NotInitializedException(String objectNotInitialized) {
-        super(exceptionMessage(objectNotInitialized));
+        super(EXCEPTION_MESSAGE);
+        this.objectNotInitialized = objectNotInitialized;
     }
 
     public NotInitializedException(String objectNotInitialized, Throwable throwable) {
```
Comment (schlosna) [inline atlasdb-commons/src/main/java/com/palantir/exception/NotInitializedException.java:40]: Should the `objectNotInitialized` be `@Safe` or `@CompileTimeConstant`

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/TargetedSweeperTest.java [hunk 3]
```diff
@@ -16,50 +16,22 @@
...
-
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.ImmutableSet;
-import com.google.common.collect.Iterables;
-import com.google.common.collect.Multimap;
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.*;
+
+import com.google.common.collect.*;
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/TargetedSweeperTest.java:28]: nit: Seems that your IDE has gone back to marking everything as * - could you please revert this? I'm surprised checkstyle didn't throw.

File: atlasdb-commons/src/main/java/com/palantir/exception/NotInitializedException.java [hunk 4]
```diff
@@ -16,18 +16,39 @@
...
-        super(exceptionMessage(objectNotInitialized), throwable);
+        super(EXCEPTION_MESSAGE, throwable);
+        this.objectNotInitialized = objectNotInitialized;
+    }
+
+    @Override
+    public @Safe String getLogMessage() {
```
Comment (mdaudali) [inline atlasdb-commons/src/main/java/com/palantir/exception/NotInitializedException.java:46]: nit: I believe it's more idiomatic to have the @Safe annotation on the return type be above the function, like with @Override.

**Review by mdaudali — COMMENTED** [review]
Thanks! If you could address the nits + @schlosna's comment on log safety, we can get this in!

File: atlasdb-commons/src/main/java/com/palantir/exception/NotInitializedException.java (hunk 2 above)
Comment (Sam-Kramer) [reply atlasdb-commons/src/main/java/com/palantir/exception/NotInitializedException.java]: ugh. have to do @safe as `getInitializingClassName()` is not considered "CompileTimeConstant"

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/TargetedSweeperTest.java [hunk 5]
```diff
@@ -230,11 +232,11 @@ public void increaseInShardsReflectedOnLastSweptTimestamp() {
     public void callingEnqueueAndSweepOnUninitializedSweeperThrows() {
         TargetedSweeper uninitializedSweeper = TargetedSweeper.createUninitializedForTest(() -> 1);
         assertThatThrownBy(() -> uninitializedSweeper.enqueue(ImmutableList.of()))
+                .isInstanceOf(NotInitializedException.class);
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/TargetedSweeperTest.java:235]: consider: also making this one an assertThatLoggableExceptionThrownBy that verifies it also returns the same args (e.g., just pulling the assert into a separate method and doing smth like assertThatNotInitializedExceptionThrownBy

**Review by mdaudali — APPROVED** [review]
thanks!

File: atlasdb-commons/src/test/java/com/palantir/exception/NotInitializedExceptionTest.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package com.palantir.exception;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.palantir.logsafe.SafeArg;
+import org.junit.Test;
+
+public class NotInitializedExceptionTest {
```
Comment (mdaudali) [inline atlasdb-commons/src/test/java/com/palantir/exception/NotInitializedExceptionTest.java:24]: nit: `public final`

File: atlasdb-commons/src/main/java/com/palantir/exception/NotInitializedException.java [hunk 7]
```diff
@@ -16,18 +16,40 @@
...
+import com.palantir.logsafe.Arg;
+import com.palantir.logsafe.Safe;
 import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.SafeLoggable;
+import java.util.List;
 
-public class NotInitializedException extends AtlasDbDependencyException {
-    public NotInitializedException(String objectNotInitialized) {
-        super(exceptionMessage(objectNotInitialized));
+public final class NotInitializedException extends AtlasDbDependencyException implements SafeLoggable {
```
Comment (mdaudali) [inline atlasdb-commons/src/main/java/com/palantir/exception/NotInitializedException.java:25]: Have you verified that this is safe to make final? (that is, there's no one who subclasses NIE) - this is on the public API so it's not necessarily safe to do this

**Review by mdaudali — CHANGES_REQUESTED** [review]

**Review by mdaudali — APPROVED** [review]

**Pull request merged**